cloud/aws/actuators/machine/actuator.go

```diff
@@ -183,6 +243,408 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+// `deleted` map.
+//
+// Second
+// First, we create a map of tags that have been created or updated by comparing
+// the `src` map to the `annotation`. If entries in the `src` are not in the
+// `annotation`, they are new. If entries in the `src` are in the `annotation`,
+// we compare their values to see if they need updating. If action must be
+// taken, entries are added to the `created` map.
+// We also build up the `newAnnotation` in this phase while building the
+// `create` map.
+//
+// This method is significantly simpler than the securityGroupsChanged method as
+// we only have to compare the `src` to our `annotation`.
+func (a *Actuator) tagsChanged(annotation map[string]interface{}, src map[string]string) (bool, map[string]string, map[string]string, map[string]interface{}) {
+	// Bool tracking if we found any changed state.
+	changed := false
+
+	// Tracking for created/updated
+	created := map[string]string{}
+
+	// Tracking for tags that were deleted.
+	deleted := map[string]string{}
+
+	// The new annotation that we need to set if anything is created/updated.
+	newAnnotation := map[string]interface{}{}
+
+	// Loop over annotation, checking if entries are in src.
+	// If an entry is present in annotation but not src, it has been deleted
+	// since last time. We flag this in the deleted map.
+	for t, v := range annotation {
+		_, ok := src[t]
+
+		// Entry isn't in src, it has been deleted.
+		if !ok {
+			// Cast v to a string here. This should be fine, tags are always
+			// strings.
+			deleted[t] = v.(string)
+			changed = true
+		}
+	}
+
+	// Loop over src, checking for entries in annotation.
+	//
+	// If an entry is in src, but not annotation, it has been created since
+	// last time.
+	//
+	// If an entry is in both src and annotation, we compare their values, if
+	// the value in src differs from that in annotation, the tag has been
+	// updated since last time.
+	//
+	for t, v := range src {
+		av, ok := annotation[t]
+
+		// Entries in the src always need to be noted in the newAnnotation. We
+		// know they're going to be created or updated.
+		newAnnotation[t] = v
+
+		// Entry isn't in annotation, it's new.
+		if !ok {
+			created[t] = v
+			newAnnotation[t] = v
+			changed = true
+			continue
+		}
+
+		// Entry is in annotation, has the value changed?
+		if v != av {
+			created[t] = v
+			changed = true
+		}
+
+		// Entry existed in both src and annotation, and their values were
+		// equal. Nothing to do.
+	}
+
+	// We made it through the loop, and everything that was in src, was also
+	// in dst. Nothing changed.
+	return changed, created, deleted, newAnnotation
+}
+
+// Returns a map[string]interface from a JSON annotation.
+// This method gets the given `annotation` from the `machine` and unmarshalls it
+// from a JSON string into a `map[string]interface{}`.
+func (a *Actuator) machineAnnotationJSON(machine *clusterv1.Machine, annotation string) (map[string]interface{}, error) {
+	u := unstructured.Unstructured{}
```

Comment:
I don't think we need unstructured in here. This is unmarshalling json into a map[string]interface{} and we could do without the extra dependency

```go
func (a *Actuator) machineAnnotationJSON(machine *clusterv1.Machine, annotation string) (map[string]interface{}, error) {
	out := map[string]interface{}{}
	jsonAnnotation := machine.GetAnnotations()[annotation]

	if len(jsonAnnotation) == 0 {
		return out, nil
	}

	// We have some kind of string, attempt to unmarshal it.
	err := json.Unmarshal([]byte(jsonAnnotation), &out)
	return out, err
}
```